Incremental draft

treblereel opened the pull request:
TODO: add details

mdproctor commented on build-caching/src/main/java/com/vertispan/j2cl/build/DiskCache.java:
```diff
@@ -436,9 +469,14 @@ private WatchKey registerWatchCreate(Path taskDir) throws IOException {
+            TaskOutput output = makeOutput(successfulResult.taskDir);
+            this.lastSuccessfulTaskDir.put(input, successfulResult.taskDir);
+            this.knownOutputs.put(successfulResult.taskDir, output);
+            if (input.getOutputType().equals(OutputTypes.TRANSPILED_JS)) {
```
I hard coded this here, as a hack. imho the writeFilesDat should be in it's own leaf task, only called once all other tasks are done.

treblereel commented:
Added hash-based processing of changed files to avoid reprocessing of unchanged files. Hash is computed from the file's
non-private members (fields and methods) at this moment. This is not a perfect solution as it does not take 
extends and implements into account, but it's a good start. So during the change detection phase, we compute the hash
of each changed file and compare it to the hash of the file in `.build.map`. If the hashes are the same, we skip deps. 

I also moved `.build.map` file writing from Turbine to Bytecode task, because javassist allows us to get the full list 
of references for the class, which is what we need to compute the deps set.

@mdproctor 
@niloc132
